pkg/ring/ring.go

```diff
@@ -396,3 +397,56 @@ func (r *Ring) Collect(ch chan<- prometheus.Metric) {
 		r.name,
 	)
 }
+
+// Subring returns a ring of n ingesters from the given ring
+// Subrings are meant only for ingestor lookup and should have their data externalized.
+func (r *Ring) Subring(key uint32, n int) (*Ring, error) {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+	if r.ringDesc == nil || len(r.ringDesc.Tokens) == 0 || n <= 0 {
+		return &Ring{}, ErrEmptyRing
+	}
+
+	if n > len(r.ringDesc.Ingesters) {
+		return nil, fmt.Errorf("subring too large")
+	}
+
+	var (
+		ingesters     = make(map[string]IngesterDesc, n)
+		distinctHosts = map[string]struct{}{}
+		start         = r.search(key)
+		iterations    = 0
+	)
+	for i := start; len(distinctHosts) < n && iterations < len(r.ringDesc.Tokens); i++ {
+		iterations++
+		// Wrap i around in the ring.
+		i %= len(r.ringDesc.Tokens)
+
+		// We want n *distinct* ingesters.
+		token := r.ringDesc.Tokens[i]
+		if _, ok := distinctHosts[token.Ingester]; ok {
+			continue
+		}
+		distinctHosts[token.Ingester] = struct{}{}
+		ingester := r.ringDesc.Ingesters[token.Ingester]
+
+		ingesters[token.Ingester] = ingester
+	}
+
+	if n > len(ingesters) {
+		return nil, fmt.Errorf("too few ingesters found")
+	}
+
+	d := &Desc{
+		Ingesters: ingesters,
+	}
+	d.Tokens = migrateRing(d) // TODO THOR there is likely a better way to do this
```

Comment:
Could you explain why is this required? I haven't deeply checked, but this function should already work on a `TokenDesc` which have already been migrated.

Reply:
I've updated this piece of code, and no longer use migrate, but instead walk the tokens list and pick out the tokens for the ingesters in our subring. That what the migrate function was doing for us thought it was a little heavy handed for our use case.